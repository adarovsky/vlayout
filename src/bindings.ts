import { Engine } from 'engine';
import { View } from 'view';
import { Button, TextField } from 'primitives';

export class Bindings {
    constructor(public readonly engine: Engine) {

    }

    registerSlot(name: string, type: string): void {
        if (type === 'listButton' && this.engine.listButtonForKey(name))
            return;

        if (type === 'listTextField' && this.engine.listTextFieldForKey(name))
            return;

        if (type === 'listView' && this.engine.listViewForKey(name))
            return;

        const v = this.viewForKey(name);
        if (!v) {
            throw new Error(`view ${name} is not registered`);
        }

        if (type === 'button' && !(v instanceof Button)) {
            throw new Error(`view ${name} is not a button, but ${v.viewType()}`);
        }

        if (type === 'textField' && !(v instanceof TextField)) {
            throw new Error(`view ${name} is not a text field, but ${v.viewType()}`);
        }
    }

    viewForKey(key: string): View | null {
        return this.engine.viewForKey(key);
    }

    toString(): string {
        return 'bindings {\n}\n';
    }
}
