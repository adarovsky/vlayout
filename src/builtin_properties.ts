import { TypeDefinition } from './types';
import { EMPTY, Observable } from 'rxjs';
import { Expression } from './expression';

export class BuiltinProperty extends Expression {
    constructor(readonly keyPath: string, readonly typeDefinition: TypeDefinition, readonly sink: Observable<any>) {
        super(0, 0);
    }

    instantiate(): this {
        const v = new (this.constructor as typeof BuiltinProperty)(this.keyPath, this.typeDefinition, this.sink);
        return v as this;
    }

    toString(): string {
        return this.keyPath;
    }
}
