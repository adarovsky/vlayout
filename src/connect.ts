import React, { createElement } from 'react';
import { combineLatest, Observable, of, Subscription } from 'rxjs';
import { fromPairs, toPairs, unzip, zip } from 'lodash';

export type ObservableMapObject<S = any> = {
    [K in keyof S]: Observable<S[K]>|S[K]
}

export function connect<P, S = {}>(WrappedComponent: React.FC<P> | React.ComponentClass<P, S>) {
    return class Connector extends React.PureComponent<ObservableMapObject<P>, P> {
        subscription = new Subscription();

        componentDidMount(): void {
            this.subscription.unsubscribe();
            const pairs = toPairs(this.props);
            const [keys, values] = unzip(pairs);
            this.subscription = combineLatest(values.map(v => v instanceof Observable ? v : of(v))).subscribe(s => {
                const pairs = zip(keys, s);
                const state = fromPairs(pairs);
                this.setState(state as P);
            });
        }

        componentWillUnmount(): void {
            this.subscription.unsubscribe();
        }

        render(): React.ReactElement<any, string | React.JSXElementConstructor<any>> | string | number | {} | React.ReactNodeArray | React.ReactPortal | boolean | null | undefined {
            if (this.state) {
                return createElement(WrappedComponent, this.state);
            } else {
                return null;
            }
        }
    }
}
