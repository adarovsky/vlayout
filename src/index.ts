export {Engine} from "./engine";
export {Layout} from "./layout";
export {ReactView, ReactViewState} from "./react_views";
import './vlayout.css';
