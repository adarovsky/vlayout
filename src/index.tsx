export {
    resizeObserver,
    pauseObserving,
    resumeObserving,
} from './resize_sensor';
export { Engine } from './engine';
export { Layout } from './layout';
import './vlayout.css';

export { ReactViewReference as ReactView } from './react_views';
export type { ReactViewProps, ReactViewState } from './react_views';
export { View } from './view';
export { ColorContainer as Color, ImageContainer as Image } from './types';
export type {
    SimpleListDefinitionItem,
    ListDefinitionItem,
    ComplexDefinitionItem,
} from './types';
export type { ListModelItem } from './list';
export { LayoutComponent } from './layout_component';
export { connect } from './connect';
export type { ObservableMapObject } from './connect';
