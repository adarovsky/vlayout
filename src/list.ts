import {AbsoluteLayout, LinearLayoutAxis, View} from "./view";
import {Layout, Scope} from "./layout";
import {LexIdentifier} from "./lexer";
import {Expression, Variable} from "./expression";
import {Dictionary, ListDefinition, TypeDefinition} from "./types";
import {Engine} from "./engine";
import {FunctionImplementationI} from "./builtin_functions";
import {LinkError} from "./errors";
import {BehaviorSubject, EMPTY, Observable, of} from "rxjs";
import _ from "lodash";
import {switchMap} from "rxjs/operators";
import {createElement} from "react";
import {ReactHorizontalList, ReactListItemPrototype, ReactVerticalList} from "./react_list";
import uuid from "uuid";

export interface ListModelItem extends Dictionary<any> {
    id: string;
}

class ListItemAccessor extends Expression {
    private modelSubject = new BehaviorSubject<ListModelItem|null>(null);
    constructor(readonly keyPath: string) {
        super(0, 0);
        this.sink = this.modelSubject.pipe(
            switchMap(m => {
                if (m === null) {
                    return EMPTY;
                }
                const prop = _.get(m, this.keyPath);
                if (prop === undefined) {
                    return EMPTY;
                }

                if (prop instanceof Observable) {
                    return prop;
                }

                return of(prop);
            })
        )
    }

    instantiate(): this {
        const v = new (this.constructor as typeof ListItemAccessor)(this.keyPath);
        return v as this;
    }

    setModelItem(modelItem: ListModelItem|null) {
        this.modelSubject.next(modelItem);
    }

    toString(): string {
        return `<model>.${this.keyPath}`;
    }
}

export class ListItemPrototype extends AbsoluteLayout implements Scope {

    accessors: Dictionary<ListItemAccessor> = {};
    modelItem: ListModelItem|null = null;

    constructor(readonly  name: LexIdentifier, readonly layout: Layout) {
        super();
        this.line = name.line;
        this.column = name.column;
    }

    private buildAccessors(source: Observable<any>, structure: Dictionary<any>, prefix: string): void {
        _.forIn(structure, (value, key) => {
            const path = prefix.length > 0 ? prefix + '.' + key : key;
            if (value instanceof TypeDefinition) {
                this.accessors[path] = new ListItemAccessor(path);
            }
            else {
                this.buildAccessors(source, value, path);
            }
        });
    }

    linkPrototype(scope: Scope, model: Variable): void {
        const modelItem = (model.typeDefinition! as ListDefinition).values[this.name.content];
        if (modelItem) {
            this.buildAccessors(model.sink, modelItem, '');
            super.link(this);
        }
        else {
            throw new LinkError(this.line, this.column, `cannot find prototype ${this.name} in model ${model.typeDefinition!.typeName}`);
        }
    }

    instantiate(): this {
        const v = new (this.constructor as typeof ListItemPrototype)(this.name, this.layout);
        v.copyFrom(this);
        return v as this;
    }

    copyFrom(what: this): void {
        super.copyFrom(what);
        this.accessors = {};
        _.forIn(what.accessors, a => this.accessors[a.keyPath] = a.instantiate());
    }

    linkModel(): void {
        super.link(this);
    }

    get engine(): Engine {
        return this.layout.engine;
    }

    functionFor(name: string, parameters: TypeDefinition[]): FunctionImplementationI {
        return this.layout.functionFor(name, parameters);
    }

    functionsLoose(name: string, parametersCount: number): FunctionImplementationI[] {
        return this.layout.functionsLoose(name, parametersCount);
    }

    variableForKeyPath(keyPath: string): Expression | null {
        return this.accessors[keyPath] || this.layout.variableForKeyPath(keyPath);
    }

    setModelItem(modelItem: ListModelItem|null): void {
        this._key = uuid.v1();
        this.modelItem = modelItem;
        if (modelItem)
            this._key = modelItem.id;
        _.forIn(this.accessors, a => a.setModelItem(modelItem));
    }

    viewType(): string {
        return `listPrototype_${this.name.content}`;
    }

    get target(): React.ReactElement<any, string | React.JSXElementConstructor<any>> {
        return createElement(ReactListItemPrototype, {parentView: this, key: this.key});
    }
}

export type ListClickHandler = (i: ListModelItem) => Promise<void>;

export class List extends View {
    model: Variable|null = null;
    tapHandler: Variable|null = null;
    tapCallback: ListClickHandler|null = null;
    prototypes: ListItemPrototype[] = [];
    private readonly reusableItems: Dictionary<ListItemPrototype[]> = {};

    constructor(readonly axis: LinearLayoutAxis) {
        super();
    }

    instantiate(): this {
        const v = new (this.constructor as typeof List)(this.axis);
        v.copyFrom(this);
        return v as this;
    }

    link(scope: Scope): void {
        super.link(scope);
        if (this.model) {
            this.model.link(scope, null);
            const t = this.model.typeDefinition;
            if (t && t instanceof ListDefinition) {
                this.prototypes.forEach(p => p.linkPrototype(scope, this.model!));
            } else if (t) {
                throw new LinkError(this.line, this.column, `model type '${t.typeName}' is not a model list`);
            }
            else {
                throw new LinkError(this.line, this.column, `model should be set for list to work`);
            }
        }
        if (this.tapHandler) {
            if (scope.engine.listButtonForKey(this.tapHandler.keyPath)) {
                this.tapCallback = scope.engine.listButtonForKey(this.tapHandler.keyPath);
            }
            else {
                throw new LinkError(this.line, this.column, `list button '${this.tapHandler.keyPath}' is not found`);
            }
        }
    }

    private createNewReusableItem(modelItem: ListModelItem): ListItemPrototype {
        const [key, value] = _.toPairs(modelItem)[0];
        const proto = this.prototypes.find(p => p.name.content === key);
        if (!proto) {
            throw new LinkError(this.line, this.column, `model item ${key} not found`);
        }

        const real = proto.instantiate(); //_.cloneDeep(proto);
        real.parent = this;
        real.linkModel();
        return real;
    }

    requestReusableItem(modelItem: ListModelItem): ListItemPrototype {
        const [key, value] = _.toPairs(modelItem)[0];
        let item: ListItemPrototype|null = null;
        if (this.reusableItems[key] &&  this.reusableItems[key].length > 0) {
            item = this.reusableItems[key].splice(0, 1)[0];
        }
        else {
            item = this.createNewReusableItem(modelItem);
        }
        item.setModelItem(value);
        return item;
    }

    returnReusableItem(proto: ListItemPrototype) {
        if (!this.reusableItems[proto.name.content]) {
            this.reusableItems[proto.name.content] = [];
        }
        this.reusableItems[proto.name.content].push(proto);
        proto.setModelItem(null);
    }


    get target(): React.ReactElement<any, string | React.JSXElementConstructor<any>> {
        return createElement(this.axis == LinearLayoutAxis.Horizontal ? ReactHorizontalList : ReactVerticalList, {parentView: this, key: this.key});
    }

    viewType(): string {
        return (this.axis == LinearLayoutAxis.Horizontal ? 'horizontal' : 'vertical') + 'List';
    }
}