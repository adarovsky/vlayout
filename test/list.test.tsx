import {mount, shallow} from "enzyme";
import {Engine, Layout} from "../src";
import React from "react";
import {BehaviorSubject, of, ReplaySubject, Subject} from "rxjs";
import {List} from "../src/list";
import sinon from "sinon";
import {SampleListView} from "./sample_list_view";
import {ElementSize} from "../src/resize_sensor";

let module = require("../src/resize_sensor");

let engine = new Engine();
const sizeChange = new Subject<ElementSize>();
let sandbox = sinon.createSandbox();

beforeEach(() => {
    engine = new Engine();
    sandbox = sinon.createSandbox();
    sandbox.stub(module, 'resizeObserver').returns(sizeChange);
});

afterEach(() => {
    sandbox.restore();
});

function getContent(type: string) {
    return `
  types {
      MyItems: list (
          user {
              name: String
          },
          newUser {
              // no fields required
          }
      )
  }
  
  inputs {
      items: MyItems
  }
  
  layout {
      layer {
          ${type} {
              padding { left: 10 right: 10 top: 10 bottom: 10 }
              model: items
  
              user {
                  label {
                      text: name
                  }
              }
              newUser {
                  label {
                      text: "add new"
                  }
              }
          }
      }
  }`;
}

describe("lists", () => {
    it("should initialize and render vertical list", async function () {
        engine.registerList("MyItems", {
            user: {
                name: engine.stringType()
            },
            newUser: {}
        });
        engine.registerInput(
            "items",
            engine.type("MyItems")!,
            of([
                {user: {id: 1, name: "Alex"}},
                {user: {id: 2, name: "Anton"}},
                {user: {id: 3, name: "Denis"}},
                {newUser: {id: "new"}}
            ])
        );

        const wrapper = mount(
            <Layout
                engine={engine}
                content={getContent('verticalList')}
            />
        );

        const node = wrapper.find(".vlayout_verticalList");

        expect(node.getDOMNode()).toMatchSnapshot();
    });

    it("should initialize and render absolute list", async function () {
        engine.registerList("MyItems", {
            user: {
                name: engine.stringType()
            },
            newUser: {}
        });
        engine.registerInput(
            "items",
            engine.type("MyItems")!,
            of([
                {user: {id: 1, name: "Alex"}},
                {user: {id: 2, name: "Anton"}},
                {user: {id: 3, name: "Denis"}},
                {newUser: {id: "new"}}
            ])
        );

        const wrapper = mount(
            <Layout
                engine={engine}
                content={getContent('absoluteList')}
            />
        );

        const node = wrapper.find(".vlayout_absoluteList");

        expect(node.getDOMNode()).toMatchSnapshot();
    });

    it("model description should support multiple properties", async function () {
        engine.registerList("MyItems", {
            user: {
                name: engine.stringType(),
                prop2: engine.boolType(),
            },
            newUser: {}
        });
        engine.registerInput(
            "items",
            engine.type("MyItems")!,
            of([
                {user: {id: 1, name: "Alex", prop2: false}},
                {user: {id: 2, name: "Anton", prop2: true}},
                {user: {id: 3, name: "Denis", prop2: false}},
                {newUser: {id: "new"}}
            ])
        );

        const wrapper = mount(
            <Layout
                engine={engine}
                content={`
                      types {
                          MyItems: list (
                              user {
                                  name: String
                                  prop2: Bool
                              },
                              newUser {
                                  // no fields required
                              }
                          )
                      }
                      
                      inputs {
                          items: MyItems
                      }
                      
                      layout {
                          layer {
                              verticalList {
                                  padding { left: 10 right: 10 top: 10 bottom: 10 }
                                  model: items
                      
                                  user {
                                      label {
                                          text: name + (prop2 ? " - good" : " - bad")
                                      }
                                  }
                                  newUser {
                                      label {
                                          text: "add new"
                                      }
                                  }
                              }
                          }
                      }`}
            />
        );

        const node = wrapper.find(".vlayout_verticalList");

        expect(node.getDOMNode()).toMatchSnapshot();
    });

    it("should initialize and render horizontal list", async function () {
        engine.registerList("MyItems", {
            user: {
                name: engine.stringType()
            },
            newUser: {}
        });
        engine.registerInput(
            "items",
            engine.type("MyItems")!,
            of([
                {user: {id: 1, name: "Alex"}},
                {user: {id: 2, name: "Anton"}},
                {user: {id: 3, name: "Denis"}},
                {newUser: {id: "new"}}
            ])
        );

        const wrapper = mount(
            <Layout
                engine={engine}
                content={getContent('horizontalList')}
            />
        );

        const node = wrapper.find(".vlayout_horizontalList");

        expect(node.getDOMNode()).toMatchSnapshot();
    });

    it("should reuse items on updates", async function () {
        // @ts-ignore
        const spy = sinon.spy(List.prototype, "createNewReusableItem");
        engine.registerList("MyItems", {
            user: {
                name: engine.stringType()
            },
            newUser: {}
        });

        let items = [
            {user: {id: 1, name: "Alex"}}
        ];
        const subj = new BehaviorSubject(items);

        engine.registerInput(
            "items",
            engine.type("MyItems")!,
            subj
        );

        const wrapper = mount(
            <Layout
                engine={engine}
                content={getContent('verticalList')}
            />
        );

        const node = wrapper.find(".vlayout_verticalList");

        expect(spy.callCount).toBe(1);
        expect(node.getDOMNode()).toMatchSnapshot();

        items = items.concat([{user: {id: 2, name: "Anton"}}]);
        subj.next(items);

        expect(spy.callCount).toBe(2);
        expect(node.getDOMNode()).toMatchSnapshot();

        spy.restore();
    });

    it("should fetch values which are observables", async function () {
        engine.registerList("MyItems", {
            user: {
                name: engine.stringType()
            },
            newUser: {}
        });

        const n = new BehaviorSubject('Anton');
        engine.registerInput(
            "items",
            engine.type("MyItems")!,
            of([
                {user: {id: 1, name: "Alex"}},
                {user: {id: 2, name: n}},
                {user: {id: 3, name: "Denis"}},
                {newUser: {id: "new"}}
            ])
        );

        const wrapper = mount(
            <Layout
                engine={engine}
                content={getContent('verticalList')}
            />
        );

        const node = wrapper.find(".vlayout_verticalList");

        expect(node.getDOMNode()).toMatchSnapshot();

        n.next('Michael');

        expect(node.getDOMNode()).toMatchSnapshot();
    });

    it("should allow setting fixed width or height for items", async function () {
        engine.registerList("MyItems", {
            user: {
                name: engine.stringType()
            },
            newUser: {}
        });

        engine.registerInput(
            "items",
            engine.type("MyItems")!,
            of([
                {user: {id: 1, name: "Alex"}},
                {user: {id: 2, name: "Anton"}},
                {user: {id: 3, name: "Denis"}},
                {newUser: {id: "new"}}
            ])
        );

        const wrapper = mount(
            <Layout
                engine={engine}
                content={`
  types {
      MyItems: list (
          user {
              name: String
          },
          newUser {
              // no fields required
          }
      )
  }
  
  inputs {
      items: MyItems
  }
  
  layout {
      layer {
          verticalList {
              padding { left: 10 right: 10 top: 10 bottom: 10 }
              model: items
  
              user {
                  fixedSize { height: 44 }
                  label {
                      text: name
                  }
              }
              newUser {
                  label {
                      text: "add new"
                  }
              }
          }
      }
  }`}
            />
        );

        const node = wrapper.find(".vlayout_verticalList");

        expect(node.getDOMNode()).toMatchSnapshot();
    });

    it("should allow nested absolutes and stacks", async function () {
        engine.registerList("MyItems", {
            user: {
                name: engine.stringType()
            },
            newUser: {}
        });

        engine.registerInput(
            "items",
            engine.type("MyItems")!,
            of([
                {user: {id: 1, name: "Alex"}},
                {user: {id: 2, name: "Anton"}},
                {user: {id: 3, name: "Denis"}},
                {newUser: {id: "new"}}
            ])
        );

        const wrapper = mount(
            <Layout
                engine={engine}
                content={`
  types {
      MyItems: list (
          user {
              name: String
          },
          newUser {
              // no fields required
          }
      )
  }
  
  inputs {
      items: MyItems
  }
  
  layout {
      layer {
          verticalList {
              padding { left: 10 right: 10 top: 10 bottom: 10 }
              model: items
  
              user {
                  absolute {
                      stack {
                          label {
                              text: name
                          }
                      }
                  }
              }
              newUser {
                  label {
                      text: "add new"
                  }
              }
          }
      }
  }`}
            />
        );

        const node = wrapper.find(".vlayout_verticalList");

        expect(node.getDOMNode()).toMatchSnapshot();
    });

    it("should support tappable items", async function () {
        engine.registerList("MyItems", {
            user: {
                name: engine.stringType()
            },
            newUser: {}
        });

        engine.registerInput(
            "items",
            engine.type("MyItems")!,
            of([
                {user: {id: 1, name: "Alex"}},
                {user: {id: 2, name: "Anton"}},
                {user: {id: 3, name: "Denis"}},
                {newUser: {id: "new"}}
            ])
        );

        engine.registerListButton('itemTapped', async item => console.log('item', item, 'tapped'));

        const wrapper = mount(
            <Layout
                engine={engine}
                content={`
  bindings {
      itemTapped: listButton
  }
  types {
      MyItems: list (
          user {
              name: String
          },
          newUser {
              // no fields required
          }
      )
  }
  
  inputs {
      items: MyItems
  }
  
  layout {
      layer {
          verticalList {
              padding { left: 10 right: 10 top: 10 bottom: 10 }
              model: items
              itemTapped: itemTapped
  
              user {
                  fixedSize { height: 44 }
                  label {
                      text: name
                  }
              }
              newUser {
                  label {
                      text: "add new"
                  }
              }
          }
      }
  }`}
            />
        );

        const node = wrapper.find(".vlayout_verticalList");

        expect(node.getDOMNode()).toMatchSnapshot();
    });

    it("should support tappable items inside hierarchy", async function () {
        engine.registerList("MyItems", {
            user: {
                name: engine.stringType()
            },
            newUser: {}
        });

        engine.registerInput(
            "items",
            engine.type("MyItems")!,
            of([
                {user: {id: 1, name: "Alex"}},
                {user: {id: 2, name: "Anton"}},
                {user: {id: 3, name: "Denis"}},
                {newUser: {id: "new"}}
            ])
        );

        engine.registerListButton('itemTapped', async item => console.log('item', item, 'tapped'));

        const wrapper = mount(
            <Layout
                engine={engine}
                content={`
  bindings {
      itemTapped: listButton
  }
  types {
      MyItems: list (
          user {
              name: String
          },
          newUser {
              // no fields required
          }
      )
  }
  
  inputs {
      items: MyItems
  }
  
  layout {
      layer {
          verticalList {
              padding { left: 10 right: 10 top: 10 bottom: 10 }
              model: items
  
              user {
                  fixedSize { height: 44 }
                  label {
                      text: name
                  }
                  absolute {
                      itemTapped {
                      }
                  }
              }
              newUser {
                  label {
                      text: "add new"
                  }
              }
          }
      }
  }`}
            />
        );

        const node = wrapper.find(".vlayout_button");

        expect(node).toHaveLength(3);
    });

    it("should support external views", async function () {
        engine.registerList("MyItems", {
            user: {
                name: engine.stringType()
            },
            newUser: {}
        });

        engine.registerInput(
            "items",
            engine.type("MyItems")!,
            of([
                {user: {id: 1, name: "Alex"}},
                {user: {id: 2, name: "Anton"}},
                {user: {id: 3, name: "Denis"}},
                {newUser: {id: "new"}}
            ])
        );

        engine.registerListView('myView', (x, item) => <SampleListView parentView={x} key={'123'} color={'#aa99cc'} user={item}/>);

        const wrapper = mount(
            <Layout
                engine={engine}
                content={`
  bindings {
      myView: listView
  }
  types {
      MyItems: list (
          user {
              name: String
          },
          newUser {
              // no fields required
          }
      )
  }
  
  inputs {
      items: MyItems
  }
  
  layout {
      layer {
          verticalList {
              padding { left: 10 right: 10 top: 10 bottom: 10 }
              model: items
  
              user {
                  fixedSize { height: 44 }
                  myView {                      
                  }
                  label {
                      text: name
                  }
              }
              newUser {
                  label {
                      text: "add new"
                  }
              }
          }
      }
  }`}
            />
        );

        const node = wrapper.find(".vlayout_verticalList");

        expect(node.getDOMNode()).toMatchSnapshot();
    });

    it("should support filtering", async function () {
        engine.registerList("MyItems", {
            user: {
                name: engine.stringType()
            },
            newUser: {}
        });

        engine.registerInput(
            "items",
            engine.type("MyItems")!,
            of([
                {user: {id: 1, name: "Alex"}},
                {user: {id: 2, name: "Anton"}},
                {user: {id: 3, name: "Denis"}},
                {newUser: {id: "new"}}
            ])
        );

        engine.registerListView('myView', (x, item) => <SampleListView parentView={x} key={'123'} color={'#aa99cc'} user={item}/>);

        const wrapper = mount(
            <Layout
                engine={engine}
                content={`
  bindings {
      myView: listView
  }
  types {
      MyItems: list (
          user {
              name: String
          },
          newUser {
              // no fields required
          }
      )
  }
  
  inputs {
      items: MyItems
  }
  
  layout {
      layer {
          verticalList {
              padding { left: 10 right: 10 top: 10 bottom: 10 }
              model: items
  
              user {
                  filter: name != "Denis"
                  fixedSize { height: 44 }
                  myView {                      
                  }
                  label {
                      text: name
                  }
              }
              newUser {
                  label {
                      text: "add new"
                  }
              }
          }
      }
  }`}
            />
        );

        const node = wrapper.find(".vlayout_verticalList");

        expect(node.getDOMNode()).toMatchSnapshot();
    });

    it("should support dynamic filtering", async function () {
        engine.registerList("MyItems", {
            user: {
                name: engine.stringType()
            },
            newUser: {}
        });

        const s = new ReplaySubject<string>(1);
        engine.registerInput(
            "items",
            engine.type("MyItems")!,
            of([
                {user: {id: 1, name: "Alex"}},
                {user: {id: 2, name: "Anton"}},
                {user: {id: 3, name: s}},
                {newUser: {id: "new"}}
            ])
        );

        const wrapper = mount(
            <Layout
                engine={engine}
                content={`
  types {
      MyItems: list (
          user {
              name: String
          },
          newUser {
              // no fields required
          }
      )
  }
  
  inputs {
      items: MyItems
  }
  
  layout {
      layer {
          verticalList {
              padding { left: 10 right: 10 top: 10 bottom: 10 }
              model: items
  
              user {
                  filter: name != "Denis"
                  fixedSize { height: 44 }
                  label {
                      text: name
                  }
              }
              newUser {
                  label {
                      text: "add new"
                  }
              }
          }
      }
  }`}
            />
        );

        const node = wrapper.find(".vlayout_verticalList");

        expect(node.getDOMNode()).toMatchSnapshot();

        s.next("Michael");
        expect(node.getDOMNode()).toMatchSnapshot();

        s.next("Denis");
        expect(node.getDOMNode()).toMatchSnapshot();
    });

    it("should support collection count", async function () {
        engine.registerList("MyItems", {
            user: {
                name: engine.stringType()
            },
            newUser: {}
        });

        const s = new ReplaySubject<string>(1);
        engine.registerInput(
            "items",
            engine.type("MyItems")!,
            of([
                {user: {id: 1, name: "Alex"}},
                {user: {id: 2, name: "Anton"}},
                {user: {id: 3, name: s}},
                {newUser: {id: "new"}}
            ])
        );

        const wrapper = mount(
            <Layout
                engine={engine}
                content={`
  types {
      MyItems: list (
          user {
              name: String
          },
          newUser {
              // no fields required
          }
      )
  }
  
  inputs {
      items: MyItems
  }
  
  layout {
      layer {
          label {
              text: String(items.count)
          }
      }
  }`}
            />
        );

        expect(wrapper.find('.vlayout_label > span').text()).toBe("4");
    });


    it("should support text field in items", async function () {
        engine.registerList("MyItems", {
            user: {
                name: engine.stringType()
            },
            newUser: {}
        });
        const spy = sandbox.spy();
        engine.registerListTextField('itemText', spy );

        const s = new BehaviorSubject<string>('Michael');
        engine.registerInput(
            "items",
            engine.type("MyItems")!,
            of([
                {user: {id: 1, name: "Alex"}},
                {user: {id: 2, name: "Anton"}},
                {user: {id: 3, name: s}},
                {newUser: {id: "new"}}
            ])
        );

        const wrapper = mount(
            <Layout
                engine={engine}
                content={`
  types {
      MyItems: list (
          user {
              name: String
          },
          newUser {
              // no fields required
          }
      )
  }
  
  inputs {
      items: MyItems
  }
  
  layout {
      layer {
          verticalList {
              padding { left: 10 right: 10 top: 10 bottom: 10 }
              model: items
  
              user {
                  itemText {
                      text: name
                  }
              }
              newUser {
                  label {
                      text: "add new"
                  }
              }
          }
      }
  }`}
            />
        );

        const node = wrapper.find(".vlayout_verticalList");

        expect(node.getDOMNode()).toMatchSnapshot();
        s.next('Sergey');

        expect(node.getDOMNode()).toMatchSnapshot();

        expect(spy.callCount).toBe(0);
        const third = node.find('.vlayout_textField').at(2);
        const input = third.find('input');
        input.simulate('change', {target: {value: 'entered text'}});

        expect(spy.callCount).toBe(1);
    });

    it("should check for model inconsistencies", async function () {
        engine.registerList("MyItems", {
            user: {
                name: engine.numberType()
            },
            newUser: {}
        });
        engine.registerInput(
            "items",
            engine.type("MyItems")!,
            of([
                {newUser: {id: "new"}}
            ])
        );

        const wrapper = () => shallow(
            <Layout
                engine={engine}
                content={`
  types {
      MyItems: list (
          user {
              name: String
          },
          newUser {
              // no fields required
          }
      )
  }
  
  inputs {
      items: MyItems
  }
  
  layout {
  }`}
            />
        );

        expect(wrapper).toThrowError(/String and Number do not match/);
    });

    it("should check for missing model", async function () {
        const wrapper = () => shallow(
            <Layout
                engine={engine}
                content={`
  types {
      MyItems: list (
          user {
              name: String
          },
          newUser {
              // no fields required
          }
      )
  }
    
  layout {
      layer {
      }
  }`}
            />
        );

        expect(wrapper).toThrowError(/list MyItems is not registered/);
    });

    it("should check for wrong model type", async function () {
        engine.registerList("MyItems", {
            user: {
                name: engine.numberType()
            },
            newUser: {}
        });
        const wrapper = () => shallow(
            <Layout
                engine={engine}
                content={`
  types {
      MyItems: enum(first, second)
  }
    
  layout {
      layer {
      }
  }`}
            />
        );

        expect(wrapper).toThrowError(/MyItems is not a enum definition/);
    });

    it("should check for wrong engine model type", async function () {
        engine.registerEnum("MyItems", {
            key1: 1, key2: 2
        });
        const wrapper = () => shallow(
            <Layout
                engine={engine}
                content={`
  types {
      MyItems: list (
          user {
              name: String
          },
          newUser {
              // no fields required
          }
      )
  }
    
  layout {
      layer {
      }
  }`}
            />
        );

        expect(wrapper).toThrowError(/MyItems is not a list definition/);
    });

    it("should check for missing type in engine model", async function () {
        engine.registerList("MyItems", {
            user: {
                name: engine.stringType()
            },
            newUser: {}
        });
        engine.registerInput(
            "items",
            engine.type("MyItems")!,
            of([
                {newUser: {id: "new"}}
            ])
        );

        const wrapper = () => shallow(
            <Layout
                engine={engine}
                content={`
  types {
      MyItems: list (
          user {
              name: String
              extraField: Number
          },
          newUser {
              // no fields required
          }
      )
  }
  
  inputs {
      items: MyItems
  }
  
  layout {
  }`}
            />
        );

        expect(wrapper).toThrowError(/extraField is missing in engine/);
    });

    it("should check for missing type in definition", async function () {
        engine.registerList("MyItems", {
            user: {
                name: engine.stringType(),
                extraField: engine.stringType()
            },
            newUser: {}
        });
        engine.registerInput(
            "items",
            engine.type("MyItems")!,
            of([
                {newUser: {id: "new"}}
            ])
        );

        const wrapper = shallow(
            <Layout
                engine={engine}
                content={`
  types {
      MyItems: list (
          user {
              name: String
          },
          newUser {
              // no fields required
          }
      )
  }

  inputs {
      items: MyItems
  }

  layout {
  }`}
            />
        );
    });

    it("should allow missing prototype in definition", async function () {
        engine.registerList("MyItems", {
            user: {
                name: engine.stringType()
            },
            newUser: {}
        });
        engine.registerInput(
            "items",
            engine.type("MyItems")!,
            of([
                {newUser: {id: "new"}}
            ])
        );

        const wrapper = shallow(
            <Layout
                engine={engine}
                content={`
  types {
      MyItems: list (
          user {
              name: String
          }
      )
  }

  inputs {
      items: MyItems
  }

  layout {
  }`}
            />
        );
    });
});
