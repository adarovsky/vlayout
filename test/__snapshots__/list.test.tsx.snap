// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lists should initialize and render vertical list 1`] = `
<div
  class="vlayout_verticalList"
  style="position: absolute; left: 10px; z-index: 1; right: 10px; top: 10px; bottom: 10px; pointer-events: auto; flex-direction: column; justify-content: stretch; display: flex; align-items: stretch; overflow-y: scroll;"
>
  <div
    class="vlayout_listPrototype_user"
    style="pointer-events: none; position: relative; min-width: 100%;"
  >
    <div
      class="vlayout_label"
      style="position: absolute; z-index: 1;"
    >
      <span>
        Alex
        <br />
      </span>
      <div
        class="vlayout_label_shadow"
        style="position: absolute; white-space: pre; opacity: 0; z-index: 1;"
      >
        <span>
          Alex
          <br />
        </span>
      </div>
    </div>
  </div>
  <div
    class="vlayout_listPrototype_user"
    style="pointer-events: none; position: relative; min-width: 100%;"
  >
    <div
      class="vlayout_label"
      style="position: absolute; z-index: 1;"
    >
      <span>
        Anton
        <br />
      </span>
      <div
        class="vlayout_label_shadow"
        style="position: absolute; white-space: pre; opacity: 0; z-index: 1;"
      >
        <span>
          Anton
          <br />
        </span>
      </div>
    </div>
  </div>
  <div
    class="vlayout_listPrototype_user"
    style="pointer-events: none; position: relative; min-width: 100%;"
  >
    <div
      class="vlayout_label"
      style="position: absolute; z-index: 1;"
    >
      <span>
        Denis
        <br />
      </span>
      <div
        class="vlayout_label_shadow"
        style="position: absolute; white-space: pre; opacity: 0; z-index: 1;"
      >
        <span>
          Denis
          <br />
        </span>
      </div>
    </div>
  </div>
  <div
    class="vlayout_listPrototype_newUser"
    style="pointer-events: none; position: relative; min-width: 100%;"
  >
    <div
      class="vlayout_label"
      style="position: absolute; z-index: 1;"
    >
      <span>
        add new
        <br />
      </span>
      <div
        class="vlayout_label_shadow"
        style="position: absolute; white-space: pre; opacity: 0; z-index: 1;"
      >
        <span>
          add new
          <br />
        </span>
      </div>
    </div>
  </div>
</div>
`;
